state/protocol/protocol_state/mutator.go

```diff
@@ -5,57 +5,197 @@ import (
+// of service events being included in a block payload:
+//   - inserting incorporated service events
+//   - updating protocol state for the candidate block
+//
+// Consider a chain where a service event is emitted during execution of block A.
+// Block B contains a receipt for A. Block C contains a seal for block A.
+//
+// A <- .. <- B(RA) <- .. <- C(SA)
+//
+// Service events are included within execution results, which are stored
+// opaquely as part of the block payload in block B. We only validate and insert
+// the typed service event to storage once we process C, the block containing the
+// seal for block A. This is because we rely on the sealing subsystem to validate
+// correctness of the service event before processing it.
+// Consequently, any change to the protocol state introduced by a service event
+// emitted during execution of block A would only become visible when querying
+// C or its descendants.
+//
+// This method will only apply service-event-induced state changes when the
+// input block has the form of block C (ie. contains a seal for a block in
+// which a service event was emitted).
+//
+// All updates that are incorporated in service events are applied to the protocol state by mutating protocol state updater.
+// This method doesn't modify any data from self, all protocol state changes are applied to state updater.
+// All other changes are returned as slice of deferred DB updates.
+//
+// Return values:
+//   - dbUpdates - If the service events are valid, or there are no service events,
+//     this method returns a slice of Badger operations to apply while storing the block.
+//     This includes operations to insert service events for blocks that include them.
+//
+// No errors are expected during normal operation.
+func (m *stateMutator) handleServiceEvents(seals []*flow.Seal) (dbUpdates []func(*transaction.Tx) error, err error) {
+	epochFallbackTriggered, err := m.params.EpochFallbackTriggered()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve epoch fallback status: %w", err)
+	}
+
+	parentProtocolState := m.stateMachine.ParentState()
+	epochStatus := parentProtocolState.EpochStatus()
+	activeSetup := parentProtocolState.CurrentEpochSetup
+
+	// never process service events after epoch fallback is triggered
+	if epochStatus.InvalidServiceEventIncorporated || epochFallbackTriggered {
+		return dbUpdates, nil
+	}
+
+	// perform protocol state transition to next epoch if next epoch is committed and we are at first block of epoch
+	phase, err := epochStatus.Phase()
+	if err != nil {
+		return nil, fmt.Errorf("could not determine epoch phase: %w", err)
+	}
+	if phase == flow.EpochPhaseCommitted {
+		if m.stateMachine.View() > activeSetup.FinalView {
+			// TODO: this is a temporary workaround to allow for the epoch transition to be triggered
+			// most likely it will be not needed when we refactor protocol state entries and define strict safety rules.
```

Comment:
I don't really understand this todo. What is going to change so that we won't need to check for epoch transition here?